### PR TITLE
[#40] Custom metrics: get column by name

### PR DESCRIPTION
**Pull request by xRay2016**
The hitogram metric is now using `get_column_by_name`.

### CODE REVIEW

File: doc/YAML.md [hunk 1]
```diff
@@ -21,4 +21,12 @@ Each yaml is sequcence of queries. Every query can include four parts. See a [sa
 |----------|---------|----------|-------------|
 | name | | No | The name of this column |
 | type | | Yes | The type of column. Valid options: `label`, `gauge`, `counter`, `histogram` |
-| description |  | No | The description of this column |
\ No newline at end of file
+| description |  | No | The description of this column |
+
+For the `histogram` type, the column name of query should be `X_sum`,  `X_count`, `X`, `X_bucket`. The `X` is the name property of histogram columns, and the order of column names should be the same. The specific meaning of the column name is as following:
+
```
Comment (jesperpedersen) [inline doc/YAML.md:27]: Maybe make this into a table as well with a number column.

```
| 1 | X_sum | The histogram sum |
```

and so on

File: doc/tutorial/02_custom_metrics.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+## Change pgexporter configuration
+
+Change `pgexporter.conf` to add 
+
+```
+metrics_path=[PATH_TO_YAML_FILE_DIR]
+```
+
+or run the command with `-Y` options.(`pgexporter` user)
```
Comment (jesperpedersen) [inline doc/tutorial/02_custom_metrics.md:25]: Missing line feed before the user description

File: pgexporter.spec [hunk 3]
```diff
@@ -47,6 +47,7 @@ cmake -DCMAKE_BUILD_TYPE=Release ..
 %{__install} -m 644 %{_builddir}/%{name}-%{version}/contrib/shell_comp/pgexporter_comp.bash %{buildroot}%{_docdir}/%{name}/shell_comp/pgexporter_comp.bash
 %{__install} -m 644 %{_builddir}/%{name}-%{version}/contrib/shell_comp/pgexporter_comp.zsh %{buildroot}%{_docdir}/%{name}/shell_comp/pgexporter_comp.zsh
 %{__install} -m 644 %{_builddir}/%{name}-%{version}/doc/tutorial/01_install.md %{buildroot}%{_docdir}/%{name}/tutorial/01_install.md
+%{__install} -m 644 %{_builddir}/%{name}-%{version}/doc/tutorial/01_install.md %{buildroot}%{_docdir}/%{name}/tutorial/02_custom_metrics.md
```
Comment (jesperpedersen) [inline pgexporter.spec:50]: The `%{_builddir}` definition is wrong

File: src/libpgexporter/queries.c [hunk 4]
```diff
@@ -505,6 +505,27 @@ pgexporter_query_debug(struct query* query)
...
+char*
+pgexporter_get_column_by_name(char* name, struct query* query, struct tuple* tuple)
+{
+   int column = -1;
+
+   for (int i = 0; i < query->number_of_columns; i++)
+   {
+      if (!strcmp(query->names[i], name))
+      {
+         column = i;
```
Comment (jesperpedersen) [inline src/libpgexporter/queries.c:517]: Put the `return` here

File: src/libpgexporter/queries.c [hunk 5]
```diff
@@ -505,6 +505,27 @@ pgexporter_query_debug(struct query* query)
...
+         break;
+      }
+   }
+
+   if (column == -1)
```
Comment (jesperpedersen) [inline src/libpgexporter/queries.c:522]: Replace with `return NULL;`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Lets see how these changes go.

However, I'm working on `RowDescription (B)` support, but that will - potentially - be merged in afterwards.

File: doc/YAML.md [hunk 6]
```diff
@@ -21,4 +21,12 @@ Each yaml is sequcence of queries. Every query can include four parts. See a [sa
 |----------|---------|----------|-------------|
 | name | | No | The name of this column |
 | type | | Yes | The type of column. Valid options: `label`, `gauge`, `counter`, `histogram` |
-| description |  | No | The description of this column |
\ No newline at end of file
+| description |  | No | The description of this column |
+
+For the `histogram` type, the column name of query should be `X_sum`,  `X_count`, `X`, `X_bucket`. The `X` is the name property of histogram columns, and the order of column names should be the same. The specific meaning of the column name is as following:
```
Comment (jesperpedersen) [inline doc/YAML.md:26]: There should be no ordering now

File: doc/YAML.md [hunk 7]
```diff
@@ -21,4 +21,12 @@ Each yaml is sequcence of queries. Every query can include four parts. See a [sa
...
+
+-   X_sum        : The histogram sum
```
Comment (jesperpedersen) [inline doc/YAML.md:28]: Should be a table

File: src/libpgexporter/prometheus.c [hunk 8]
```diff
@@ -1927,7 +1927,8 @@ histogram_information(struct prometheus* prom, int client_fd)
    int ret;
    int n_bounds = 0;
    int n_buckets = 0;
-   int histogram_start_idx;
+   int histogram_idx;
+   char* names[MAX_ARR_LENGTH];
```
Comment (jesperpedersen) [inline src/libpgexporter/prometheus.c:1931]: We don't need `names` anymore

File: src/libpgexporter/prometheus.c [hunk 9]
```diff
@@ -1969,7 +1981,7 @@ histogram_information(struct prometheus* prom, int client_fd)
             continue;
          }
 
-         ret = pgexporter_custom_query(server, prom->query, prom->tag, 5, NULL, &query);
+         ret = pgexporter_custom_query(server, prom->query, prom->tag, prom->number_of_columns + 3, names, &query);
```
Comment (jesperpedersen) [inline src/libpgexporter/prometheus.c:1984]: `-1` and `NULL` now

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by xRay2016** [comment]
OK! I will change teh code accroding to the code and will use `RowDescription (B)` support.

File: doc/tutorial/02_custom_metrics.md (hunk 2 above)
Comment (xRay2016) [reply doc/tutorial/02_custom_metrics.md:25]: I don't understand this clearly.

File: src/libpgexporter/prometheus.c (hunk 8 above)
Comment (xRay2016) [reply src/libpgexporter/prometheus.c]: I now use `names` in `get_column_by_name` only.

**Comment by jesperpedersen** [comment]
I have made some changes so go over the commit pushed; the biggest change was `_buckets{` to `_bucket{`

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request closed without merging**